### PR TITLE
add new config to batch properties calls

### PR DESCRIPTION
**Pull request by gyDBD**
## Feature
Add a new config `propertyBatchKey` so that we are able to batch “properties” call.
Only a specific kind of endpoint would benefit from this: 
 input parameters format:
```
ids: str[] (should be the batchKey)
properties: str[] (should be the propertyBatchKey)
```

response:
```
properties is not returned in a nested object, but spread at top level. (see getFilmsV2 in example/swapi)
e.g.: { id: 'id_1', property_1: 'foo', property_2: 'boo' }
``` 

## Status Quo
run `node build/swapi-server.js` under `example/swapi`, the `getFilmsV2` function will be called three times in a roll.

## after this change
run `node build/swapi-server.js` under `example/swapi`, the `getFilmsV2` function will be only be called once. 
And the flow type of response
Screen Shot 2021-07-01 at 10 40 59 AM

## Manual testing enforce optional properties 
https://fluffy.yelpcorp.com/i/18BHFxFfJg0TgsWjWDwngS2c6D4WS8zS.html#L1,L7,L12,L15,L29,L35,L57

## Note
I realized it's hard to build something perfect that works for different kinds of use cases. And it adds complexity for both reviewers and me. Given we don't have much time left for this project, I think we should only focus on to support the use case of batching SBP services calls and maybe in the future we can improve this more. 

(Previous pr: https://github.com/Yelp/dataloader-codegen/pull/261)

### CODE REVIEW

File: src/runtimeHelpers.ts [hunk 1]
```diff
@@ -253,6 +298,141 @@ export function unPartitionResults<T>(
...
+                        break;
+                    }
+
+                    if (resultElement[newKey] === batchKeyPartition[i][j]) {
+                        result = Object.assign(
+                            {},
+                            ...[newKey, ...propertyBatchKeyPartion[i][j]].map((key) => ({
+                                [key]: resultElement[key],
+                            })),
+                        );
```
Comment (magicmark) [inline src/runtimeHelpers.ts:389]: 1. the logic here is quite dense with multiple levels of spreading....can we split this up at all?
2. `Object.assign` is ok but unnecessary- you can use straight up object literal spread: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals

File: src/runtimeHelpers.ts [hunk 2]
```diff
@@ -253,6 +298,141 @@ export function unPartitionResults<T>(
...
+ * propertyBatchKey is not returned in a nested object, but spread at top level as well.
+ * If we have 'bar_id' as newKey and 'properties' as propertyBatchKey,
+ * the resultGroups should look like this:
+ * [
+ *      [ { bar_id: 2, name: 'Burger King', rating: 3 } ],
+ *      [ { bar_id: 1, name: 'In N Out', rating: 4 } ]
+ * ],
+ *
+ *
+ * IMPORTANT NOTE: The contract must have a one-to-one correspondence between the input propertyBatchKey and the output propertyBatchKey.
```
Comment (magicmark) [inline src/runtimeHelpers.ts:315]: Can we write up this contract somewhere?

File: API_DOCS.md [hunk 3]
```diff
@@ -125,6 +126,7 @@ Describes the shape and behaviour of the resources object you will pass to `getL
 | `commaSeparatedBatchKey` | (Optional) Set to true if the interface of the resource takes the batch key as a comma separated list (rather than an array of IDs, as is more common). Default: false                                                                                   |
 | `isResponseDictionary`   | (Optional) Set to true if the batch resource returns the results as a dictionary with key mapped to values (instead of a list of items). If this option is supplied `reorderResultsByKey` should not be. Default: false                                  |
 | `isBatchKeyASet`         | (Optional) Set to true if the interface of the resource takes the batch key as a set (rather than an array). For example, when using a generated clientlib based on swagger where `uniqueItems: true` is set for the batchKey parameter. Default: false. |
+| `propertyBatchKey`       | (Optional) The argument to the resource that represents the nested list of optional properties we want to fetch. (e.g. usually 'properties' or 'features').                                                                                              |
```
Comment (magicmark) [inline API_DOCS.md:129]: "represents the nested list"

but it isn't nested? the pr description says:

> properties is **not** returned in a nested object, but spread at top level. (see getFilmsV2 in example/swapi)

File: __tests__/implementation.test.js [hunk 4]
```diff
@@ -1217,3 +1217,233 @@ test('bail if errorHandler does not return an error', async () => {
...
+            }
+
+            if (_.isEqual(foo_ids, [3])) {
+                expect(include_extra_info).toBe(true);
+                return Promise.resolve([
+                    {
+                        foo_id: 3,
+                        rating: 5,
+                        name: 'Shake Shack',
+                        extra_stuff: 'lorem ipsum',
```
Comment (magicmark) [inline __tests__/implementation.test.js:1251]: is extra_stuff relevant here? does this violate the contract?

File: src/runtimeHelpers.ts (hunk 1 above)
Comment (gyDBD) [reply src/runtimeHelpers.ts]: I used `_pick` instead.

**Comment by gyDBD** [comment]
I added swagger parser and validate only reponseKey (they key in the response that correspond to batchKey, usually id) can be marked as required in `src/index.ts` and manually tested https://fluffy.yelpcorp.com/i/18BHFxFfJg0TgsWjWDwngS2c6D4WS8zS.html#L1,L7,L12,L15,L29,L35,L57

File: README.md [hunk 5]
```diff
@@ -147,6 +147,45 @@ resources:
         newKey: user_id
 ```
 
+### Batch Resources with `properties` paramerters
+
+Instead of accepting just a list of users (`user_ids`), a batch resouce could accept both a list of users (`user_ids`) and a list of properties (`properties`):
```
Comment (magicmark) [inline README.md:152]:
```suggestion
Instead of accepting just a list of users (`user_ids`), a batch resource could accept both a list of users (`user_ids`) and a list of properties (`properties`) to fetch about that user:
```

File: README.md [hunk 6]
```diff
@@ -147,6 +147,45 @@ resources:
...
+
+```js
+const getUserInfo = (args: { user_ids: Array<number>, properties: Array<string> }): Promise<Array<UserInfo>> =>
+    fetch('/userInfo', args);
```
Comment (magicmark) [inline README.md:156]:
```suggestion
const users = getUsers({
  ids: [1, 2, 3],
  properties: ['firstName', 'age'],
})

/**
 * e.g. users => 
 * [
 *   { firstName: 'Alice', age: 42 },
 *   { firstName: 'Bob', age: 70 }
 * ]
 */
```

Just to clarify with example query and importantly, response

File: README.md [hunk 7]
```diff
@@ -147,6 +147,45 @@ resources:
...
+
+1. The resouce accept two non-optional `list` type parameters, like this:
+
+```
+ids: str[] (should be the batchKey)
+properties: str[] (should be the propertyBatchKey)
+```
+
+2. `properties` are spread at the same level as the `responseKey`. (see `getFilmsV2` in [swapi example](./examples/swapi/swapi-server.js))
+3. All `properties` should be optional in the response object. To enfore that, we require `swaggerLink`, `swaggerPath` and `httpMethod` defined in the config.
```
Comment (magicmark) [inline README.md:185]:
```suggestion
3. All `properties` MUST be optional in the response object. To enfore that, we require `swaggerLink`, `swaggerPath` and `httpMethod` defined in the config.
```

https://www.ietf.org/rfc/rfc2119.txt

File: src/index.ts [hunk 8]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+            return;
+        }
+    });
+    return array;
+}
+
+/**
+ * Throw erros when resource uses propertyBatchKey but doesn't match requirements
+ */
+function verifyBatchPropertyResource(args: CLIArgs) {
```
Comment (magicmark) [inline src/index.ts:56]: As discussed, since this is quite Yelpy and relies on checking swagger stuff specifically (which this tool is currently opinionated about), I'd suggest keeping this logic but moving it to a seperate script that runs in our build, outside of this tool.

That way, we can remove the swaggerPath / httpMethod keys etc since we'll know how to map to it.

File: README.md (hunk 7 above)
Comment (magicmark) [reply README.md:185]: per https://github.com/Yelp/dataloader-codegen/pull/299/files#r675871585

I'd change "The flow types currently don't handle the nullability of these properties correctly, so to enforce this, we recommend a build step to ensure that the underlying types are always set as maybe types.

File: README.md [hunk 9]
```diff
@@ -147,6 +147,45 @@ resources:
...
+4. The resouce must have a one-to-one correspondence between the input "properties" and the output "properties".
+   (i.e. if we request property "name", the response must have "name" in it, and no extra data assciated with it.)
```
Comment (magicmark) [inline README.md:187]: nit: https://www.merriam-webster.com/words-at-play/ie-vs-eg-abbreviation-meaning-usage-difference

```suggestion
   - (e.g. if we request property "name", the response must have "name" in it, and no extra data associated with it.)
```

File: src/index.ts [hunk 10]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
     fs.writeFileSync(args.output, output);
 }
 
+/**
+ * If requiredList only contains responseKey
```
Comment (magicmark) [inline src/index.ts:26]: this is pretty abstract - what's requiredList?
(an example always helps)

File: src/index.ts [hunk 11]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+function validRequiredList(requiredList: Array<string>, responseKey: string) {
+    if (requiredList.length > 1) return false;
+    if (requiredList.length == 1 && requiredList[0] !== responseKey) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * Find all values of the intput key in an object recursively
```
Comment (magicmark) [inline src/index.ts:37]:
```suggestion
 * Find all values of the input key in an object recursively
```

would also benefit from an example, i'm not sure what this code is supposed to do by looking at it

File: src/index.ts [hunk 12]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+ */
+function findVal(object: any, key: string, array: Array<string>) {
+    Object.keys(object).some(function (k) {
+        if (k === key) {
+            array.push(...object[k]);
```
Comment (magicmark) [inline src/index.ts:42]: does eslint not complain at you? we're mutating the input array variable here, which probably isn't what you want (unintended side effects)

objects are pass by reference in js https://www.javascripttutorial.net/javascript-pass-by-value/

File: src/index.ts [hunk 13]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+                throw new Error(`Missing swagger info for ${key}, please add them in the dataloader-cofig.yaml file!`);
+            }
+            const swaggerLink: string = value['swaggerLink'];
+            const swaggerPath: string = value['swaggerPath'];
+            const httpMethod: string = value['httpMethod'];
+
+            // parse swagger file, so that all $ref pointers will be resolved.
+            SwaggerParser.validate(swaggerLink, (err, api) => {
+                if (err) {
+                    console.error(err);
```
Comment (magicmark) [inline src/index.ts:77]: is silently failing a good thing? do we have cases where the swagger spec fails to parse but "still works"? or can we hard fail here?

File: src/index.ts [hunk 14]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+                } else {
+                    if (!api || !api.paths[swaggerPath][httpMethod]) {
+                        throw new Error(
+                            `Cannot find the swagger response definition for ${key}, please make sure you have correct swagger info in the dataloader-cofig.yaml file!`,
+                        );
+                    } else {
```
Comment (magicmark) [inline src/index.ts:83]: don't need an else after throw new Error, the runtime will never get past this line if it throws

File: src/index.ts [hunk 15]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+                        var swaggerResponse = api.paths[swaggerPath][httpMethod]['responses']['200'];
```
Comment (magicmark) [inline src/index.ts:84]:
```suggestion
                        let swaggerResponse = api.paths[swaggerPath][httpMethod]['responses']['200'];
```

https://stackoverflow.com/questions/762011/whats-the-difference-between-using-let-and-var

File: src/index.ts [hunk 16]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+                        // The resource may return the list of results in a nested path and properties are under the nestedPath
+                        if (value.hasOwnProperty('nestedPath')) {
+                            swaggerResponse =
+                                api.paths[swaggerPath][httpMethod]['responses']['200']['schema'][value['nestedPath']];
+                        }
+                        // Find all the fields that are `required` in the swagger spec
+                        const requiredList = findVal(swaggerResponse, 'required', []);
```
Comment (magicmark) [inline src/index.ts:91]: is there a lodash method for this by any chance? also if it's only like 3 loc and only used here, i'd just inline the logic to avoid the layer of indirection

File: src/index.ts [hunk 17]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+                        const responseKey = value['responseKey'];
+                        if (!validRequiredList(requiredList, responseKey)) {
+                            throw new Error(
+                                [
+                                    'Sorry, your swagger endpoint does not match the requirement of using propertyBatchKey,',
```
Comment (magicmark) [inline src/index.ts:96]: which endpoint? looking at the stacktrace in the fluffy you shared, it wasn't obvious which dataloader-codegen yaml file would need fixing

Comment (magicmark) [reply src/index.ts:96]: (can we spit out the the name of the endpoint, the bad file etc)

File: API_DOCS.md [hunk 18]
```diff
@@ -125,6 +130,12 @@ Describes the shape and behaviour of the resources object you will pass to `getL
...
 | `isResponseDictionary`   | (Optional) Set to true if the batch resource returns the results as a dictionary with key mapped to values (instead of a list of items). If this option is supplied `reorderResultsByKey` should not be. Default: false                                  |
 | `isBatchKeyASet`         | (Optional) Set to true if the interface of the resource takes the batch key as a set (rather than an array). For example, when using a generated clientlib based on swagger where `uniqueItems: true` is set for the batchKey parameter. Default: false. |
+| `propertyBatchKey`       | (Optional) The argument to the resource that represents the optional properties we want to fetch. (e.g. usually 'properties' or 'features').                                                                                                             |
+| `responseKey`            | (Non-optional when propertyBatchKey is used) The key in the response objects corresponds to `batchKey`. This should be the only field that are marked as required in your swagger endpoint response, except nestedPath.                                  |
```
Comment (magicmark) [inline API_DOCS.md:134]: is this the same as `newKey`? looks like they share a similar purpose and will likely be the same values. Can we reuse (maybe rename) this?

**Review by magicmark — COMMENTED** [review]
implementation looking good!

comments are mostly around the docs and moving the script to enforce the swagger contract to Yelp-land

File: src/index.ts [hunk 19]
```diff
@@ -21,6 +22,89 @@ function writeLoaders(args: CLIArgs) {
...
+            const swaggerLink: string = value['swaggerLink'];
+            const swaggerPath: string = value['swaggerPath'];
+            const httpMethod: string = value['httpMethod'];
+
+            // parse swagger file, so that all $ref pointers will be resolved.
+            SwaggerParser.validate(swaggerLink, (err, api) => {
+                if (err) {
+                    console.error(err);
+                } else {
+                    if (!api || !api.paths[swaggerPath][httpMethod]) {
```
Comment (magicmark) [inline src/index.ts:79]: I think we can just search GET then POST rather than making folks define it - if there's a clash and both are defined.....maybe throw an error for now? (how do we currently handle that situation when generating the clientlibs? we don't namespace the generated method names with the http method...so assuming maybe this just hasn't happened yet? or isn't allowed?)

File: src/index.ts (hunk 14 above)
Comment (magicmark) [reply src/index.ts:83]: also we could print out the specific file path rather than the generic "the dataloader-cofig.yaml file" message

File: src/index.ts (hunk 8 above)
Comment (gyDBD) [reply src/index.ts:56]: Hmm, you mean move the script to ggs? But the `propertyBatchKey` needs to be in dataloader-codegen, if anyone else uses `propertyBatchKey` they won't have correct flow type.

Comment (gyDBD) [reply src/index.ts:56]: nvm, saw your other command!

File: API_DOCS.md (hunk 18 above)
Comment (gyDBD) [reply API_DOCS.md:134]: They might be the same, or completely different! 
`newKey` is the singular version of the `batchKey`.
However, `responseKey` is the key correspond to `batchKey` in the swagger response, we cannot control what it is. 

For example, we have this swagger endpoint, input ags
```
user_ids: str[] 
properties: str[] 
```
response: 
```
{
id: str,
name: str,
age: number,
....
}
```
batchKey is `user_ids` and newKey is `user_id` but responseKey is `id`, which is different from newKey.

File: README.md (hunk 5 above)
Comment (gyDBD) [reply README.md:152]: Thanks for catching the spell errors! I installed the code spell checker on VS, it will help me in the future.

File: README.md (hunk 6 above)
Comment (gyDBD) [reply README.md:156]: I keep the line 155 and 156, I think the definition of the resource is also the part of this example.

**Comment by gyDBD** [comment]
script to enforce the swagger contract will be moved to Yelp-land. The review comments on those lines will be addressed there.

File: schema.json [hunk 20]
```diff
@@ -97,6 +97,14 @@
...
                     "type": "boolean",
                     "description": "(Optional) Set to true if the interface of the resource takes the batch key as a set (rather than an array). For example, when using a generated clientlib based on swagger where `uniqueItems: true` is set for the batchKey parameter. Default: false"
+                },
+                "propertyBatchKey": {
+                    "type": "string",
+                    "description": "(Optional) The argument to the resource that represents the optional properties we want to fetch. (e.g. usually 'properties' or 'features')"
+                },
+                "responseKey": {
+                    "type": "string",
+                    "description": "(Non-optional when propertyBatchKey is used) The key in the response objects corresponds to `batchKey`. This should be the only field that are marked as required in your swagger endpoint response, except nestedPath."
```
Comment (magicmark) [inline schema.json:107]: > Non-optional when propertyBatchKey is used

Can we encode this via the schema?

Comment (magicmark) [reply schema.json:107]: (if it's too fiddly, we can throw an error in the codegen instead)

File: src/implementation.ts [hunk 21]
```diff
@@ -153,19 +156,42 @@ function getBatchLoader(resourceConfig: BatchResourceConfig, resourcePath: Reado
...
+             * partitionItems(['bar_id', 'properties'], [
+             *   { bar_id: 7, properties: ['property_1'], include_extra_info: true },
+             *   { bar_id: 8, properties: ['property_2'], include_extra_info: false },
+             *   { bar_id: 9, properties: ['property_3'], include_extra_info: true },
+             * ])
+             * \`\`\`
+             *
+             * Returns:
+             * \`[ [ 0, 2 ], [ 1 ] ]\`
              * We'll refer to each element in the group as a "request ID".
```
Comment (magicmark) [inline src/implementation.ts:184]: Maybe keep this directly after the first example

File: src/runtimeHelpers.ts [hunk 22]
```diff
@@ -253,6 +298,140 @@ export function unPartitionResults<T>(
...
+
+                    // Find the response that matches the requested batchKey
+                    if (resultElement[responseKey] === batchKeyPartition[i][j]) {
+                        // Only responseKey and the requested properties will be in the final result.
+                        result = _.pick(resultElement, [responseKey, ...propertyBatchKeyPartion[i][j]]);
+                        break;
+                    }
+                }
+                // If requested property doesn't exist in resultElement, we should throw BatchItemNotFoundError error.
+                if (result === null) {
```
Comment (magicmark) [inline src/runtimeHelpers.ts:393]:
```suggestion
                }

                // If requested property doesn't exist in resultElement, we should throw BatchItemNotFoundError error.
                if (result === null) {
```

nit: spacing between stanzas

File: src/runtimeHelpers.ts [hunk 23]
```diff
@@ -253,6 +298,140 @@ export function unPartitionResults<T>(
...
+                    return {
+                        order: id,
+                        result: new BatchItemNotFoundError(
+                            [
+                                `Could not find ${responseKey} = ${batchKeyPartition[i][j]} in the response dict.`,
+                                `Or your endpoint does not follow the contract we support.`,
```
Comment (magicmark) [inline src/runtimeHelpers.ts:399]: "Or your endpoint does not follow the contract we support" might be something a random feature team engineer reads whilst investigating an oncall incident for their page at some point - they might not be super familiar with dataloader-codegen and think "what contract is this talking about?"

might be worth adding a pointer to the docs or something in the error message to make this more actionable?

File: src/runtimeHelpers.ts [hunk 24]
```diff
@@ -253,6 +298,140 @@ export function unPartitionResults<T>(
...
+                            ].join(' '),
+                        ),
+                    };
+                } else {
+                    return { order: id, result: result };
```
Comment (magicmark) [inline src/runtimeHelpers.ts:404]:
```suggestion
                    return { order: id, result };
```

fun fact if the variable name and property name are the same you can just do this (surprised eslint didn't take care of this)

File: __tests__/implementation.test.js [hunk 25]
```diff
@@ -1217,3 +1217,235 @@ test('bail if errorHandler does not return an error', async () => {
...
+                propertyBatchKey: 'properties',
+                responseKey: 'foo_id',
+            },
+        },
+    };
+
+    const resources = {
+        foo: ({ foo_ids, bar }) => {
+            if (_.isEqual(foo_ids, [1, 2, 3])) {
+                return Promise.resolve([
```
Comment (magicmark) [inline __tests__/implementation.test.js:1421]:
```suggestion
            if (_.isEqual(foo_ids, [1, 2, 3])) {
                expect(bar).toBe(true);
                return Promise.resolve([
```

i assume?

File: __tests__/implementation.test.js [hunk 26]
```diff
@@ -1217,3 +1217,235 @@ test('bail if errorHandler does not return an error', async () => {
...
+                    { foo_id: 3, rating: 4, name: 'Shake Shack' },
+                    { foo_id: 1, rating: 3, name: 'Burger King' },
+                    // deliberately omit 2
+                ]);
+            } else if (_.isEqual(foo_ids, [4])) {
+                return Promise.resolve([{ foo_id: 4, rating: 5, name: 'In N Out' }]);
```
Comment (magicmark) [inline __tests__/implementation.test.js:1427]:
```suggestion
            } else if (_.isEqual(foo_ids, [4])) {
                expect(bar).toBe(false);
                return Promise.resolve([{ foo_id: 4, rating: 5, name: 'In N Out' }]);
```

ditto?

File: __tests__/implementation.test.js (hunk 25 above)
Comment (magicmark) [reply __tests__/implementation.test.js:1421]: I'm assuming "bar" is supposed to serve a similar purpose to include_extra_info - i.e. to make a different chunk and test multiple request groups?

File: README.md [hunk 27]
```diff
@@ -147,6 +147,57 @@ resources:
...
+        batchKey: user_ids
+        newKey: user_id
+        propertyBatchKey: properties
+        responseKey: id
+```
+
+**IMPORTANT NOTE**
+To use this feature, there are several restrictions. (Please open an issue if you're interested in helping us support other use cases):
+
+1. The resource accept two non-optional `list` type parameters, like this:
```
Comment (magicmark) [inline README.md:189]:
```suggestion
**Contract**

1. The resource accepts two non-optional `list` type parameters:
```

File: README.md [hunk 28]
```diff
@@ -147,6 +147,57 @@ resources:
         newKey: user_id
 ```
 
+### Batch Resources with `properties` paramerters
```
Comment (magicmark) [inline README.md:150]: Can we move this to API_DOCS.md? Seems more appropriate than the main readme

File: README.md (hunk 6 above)
Comment (magicmark) [reply README.md:156]: sounds good

File: README.md [hunk 29]
```diff
@@ -147,6 +147,57 @@ resources:
...
+  user_ids: [1, 2, 3],
+  properties: ['firstName', 'age']
+});
+
+e.g. users =>
+[
+  { id: 1, firstName: 'Alice', age: 42 },
+  { id: 2, firstName: 'Bob', age: 70 },
+  { id: 3, firstName: 'Carol', age: 50 },
+]
```
Comment (magicmark) [inline README.md:168]: should be surrounded by comments to avoid looking weird w/ syntax highlighting

File: README.md [hunk 30]
```diff
@@ -147,6 +147,57 @@ resources:
...
+        propertyBatchKey: properties
+        responseKey: id
+```
+
+**IMPORTANT NOTE**
+To use this feature, there are several restrictions. (Please open an issue if you're interested in helping us support other use cases):
+
+1. The resource accept two non-optional `list` type parameters, like this:
+
+```
```
Comment (magicmark) [inline README.md:191]: You can preview what this looks like here: https://github.com/Yelp/dataloader-codegen/blob/344462dc8d534f5f3510b6cff4e46bf87bd8784f/README.md#batch-resources-with-properties-paramerters (or there's vscode plugins to preview the rendered markdown live)

this probably should be indented inline with the above bullet

File: README.md [hunk 31]
```diff
@@ -147,6 +147,57 @@ resources:
...
+ids: str[] (should be the batchKey)
+properties: str[] (should be the propertyBatchKey)
```
Comment (magicmark) [inline README.md:193]: This feels slightly ambiguous - does the function accept two arguments, the first id and the second properties? or an object with these properties? (I know it's the second...I think?) probably worth clarifying tho

Comment (magicmark) [reply README.md:193]: maybe something similar to how we document other fns: https://github.com/Yelp/dataloader-codegen/blob/344462dc8d534f5f3510b6cff4e46bf87bd8784f/API_DOCS.md#getloaders-arguments

```js
({
  // this is the batchKey
  ids: Array<string>,
  // this is the propertyBatchKey
  properties: Array<string>,
}): Array<T>

File: README.md [hunk 32]
```diff
@@ -147,6 +147,57 @@ resources:
...
+```
+
+2. `properties` are spread at the same level as the `responseKey`. (see `getFilmsV2` in [swapi example](./examples/swapi/swapi-server.js))
```
Comment (magicmark) [inline README.md:196]:
```suggestion
2. In the response, `properties` are spread at the same level as the `responseKey`. (see `getFilmsV2` in [swapi example](./examples/swapi/swapi-server.js))
```

**Review by magicmark — APPROVED** [review]
fix n ship, awesome stuff! 🎉🎉🎉

**Review by magicmark — APPROVED** [review]
fix n ship, awesome stuff! 🎉🎉🎉

File: src/implementation.ts (hunk 21 above)
Comment (magicmark) [reply src/implementation.ts:184]: or if not, keep the spacing between the return example and this note

File: schema.json (hunk 20 above)
Comment (gyDBD) [reply schema.json:107]: Yes, we can encode this via the schema! https://stackoverflow.com/questions/38717933/jsonschema-attribute-conditionally-required (want to share with you in case you are interested)

File: README.md (hunk 28 above)
Comment (gyDBD) [reply README.md:150]: The reason I added it here is, we have `## Batch Resources` in README and `Batch Resources with properties paramerters`  is a feature on the top of batch resources, they are somewhat related. While, API_DOCS is more about how to write a config file and the documentation of each config. 

I think it's okay to leave it in README, but if you think it will make the README too wordy or maybe we don't want to advertise this feature, I could move it to API_DCOS. Please let me know.

Comment (magicmark) [reply README.md:150]: either API_DOCS.md or could be a seperate file even - this this feels like an important but "power user" feature that doesn't need to be on the homepage

File: schema.json (hunk 20 above)
Comment (magicmark) [reply schema.json:107]: cool cool - so yeah let's enforce that when propertyBatchKey is present, responseKey is also present too

File: README.md (hunk 31 above)
Comment (gyDBD) [reply README.md:193]: In the swagger spec, the endpoint takes 2 arguments. But if we are talking about the dataloader arguments, it takes an object like this: 
```
({
  // this is the batchKey
  ids: Array<string>,
  // this is the propertyBatchKey
  properties: Array<string>,
}): Array<T>
```
I think we want restrictions on swagger spec here. Any more thoughts?

Comment (magicmark) [reply README.md]: Not sure I understand the question here?

Comment (gyDBD) [reply README.md]: My question: whether we should put (the argument definition from the swagger spec)
```
ids: str[] (should be the batchKey)
properties: str[] (should be the propertyBatchKey)
``` 
vs what you have here 
```
({
  // this is the batchKey
  ids: Array<string>,
  // this is the propertyBatchKey
  properties: Array<string>,
}): Array<T>
```

Sorry, I am little confused what's `the function` and `the resources` you are referring to.

Comment (gyDBD) [reply README.md]: nvm, I think I get what you mean.

**Comment by gyDBD** [comment]
@magicmark  Hey Mark, thanks so much for the review!! I've addressed all your comments,do you want to take a final look?

**Review by magicmark — APPROVED** [review]
🥳 🎉
